src/gen_rpc_auth.erl

```diff
@@ -0,0 +1,542 @@
+%% distributed under the License is distributed on an "AS IS" BASIS,
+%% WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+%% See the License for the specific language governing permissions and
+%% limitations under the License.
+%%--------------------------------------------------------------------
+-module(gen_rpc_auth).
+
+-include("logger.hrl").
+-include_lib("snabbkaffe/include/trace.hrl").
+
+-ifdef(TEST).
+-include_lib("proper/include/proper.hrl").
+-include_lib("eunit/include/eunit.hrl").
+-endif.
+
+%% API:
+-export([connect_with_auth/3, authenticate_client/3, get_cookie/0]).
+
+-export_type([auth_error/0]).
+
+%%================================================================================
+%% Types
+%%================================================================================
+
+-type secret() :: binary().
+
+-type challenge() :: binary().
+
+-type packet() :: binary().
+
+%% Packets
+-record(gen_rpc_authenticate_c,
+        { challenge :: binary()
+        }).
+
+-record(gen_rpc_authenticate_cr,
+        { response  :: binary()
+        , challenge :: binary()
+        }).
+
+-record(gen_rpc_authenticate_r,
+        { response  :: binary()
+        }).
+
+-define(UNAUTHORIZED, {badrpc, invalid_cookie}).
+
+-define(BADPACKET, {badrpc, invalid_message}).
+
+-type auth_error() :: {badtcp, _}          %% Network errors
+                    | ?BADPACKET           %% Bad packet
+                    | ?UNAUTHORIZED.       %% Bad cookie
+
+%%================================================================================
+%% API
+%%================================================================================
+
+-spec connect_with_auth(module(), node(), inet:port()) ->
+                {ok, _Socket}
+              | {error, auth_error()}
+              | {unreachable, _Reason}.
+connect_with_auth(Driver, Node, Port) ->
+    Fallback = insecure_fallback(),
+    case connect_with_auth(Driver, Node, Port, fun authenticate_server_cr/2) of
+        {ok, Socket} ->
+            {ok, Socket};
+        Err when Fallback, Err =:= {error, {badtcp, closed}} orelse
+                           Err =:= {error, ?UNAUTHORIZED} ->
+            ?tp(warning, gen_rpc_insecure_fallback, #{peer => {Node, Port}, role => client}),
+            connect_with_auth(Driver, Node, Port, fun authenticate_server_insecure/2);
+        Result ->
+            Result
+    end.
+
+-spec authenticate_client(module(), term(), tuple()) -> ok | {error, auth_error()}.
+authenticate_client(Driver, Socket, Peer) ->
+    ok = Driver:set_send_timeout(Socket, gen_rpc_helper:get_send_timeout(undefined)),
+    RecvTimeout = gen_rpc_helper:get_authentication_timeout(),
+    Fallback = insecure_fallback(),
+    case Driver:recv(Socket, 0, RecvTimeout) of
+        {ok, Data} ->
+            case authenticate_client_cr(Driver, Socket, Data) of
+                {error, ?BADPACKET} when Fallback ->
```

Comment:
it's secure enough to always fallback at stage2?

Reply:
Could you elaborate?